### ISSUE TITLE
v2.0.0-alpha4: error in SSM middleware

### ISSUE THREAD

**Issue by theburningmonk** [middyjs/middy#611]
Since upgrading to v2.0.0-alpha4, a few of the students in my workshop have reported seeing this error during their exercises:

```
{
    "errorType": "SyntaxError",
    "errorMessage": "Unexpected token u in JSON at position 0",
    "stack": [
        "SyntaxError: Unexpected token u in JSON at position 0",
        "    at JSON.parse (<anonymous>)",
        "    at Object.onChange (/var/task/functions/get-restaurants.js:36:25)",
        "    at ssmMiddlewareBefore (/var/task/node_modules/@middy/ssm/index.js:157:175)",
        "    at runMiddlewares (/var/task/node_modules/@middy/core/index.js:207:86)",
        "    at middyPromise (/var/task/node_modules/@middy/core/index.js:100:15)",
        "    at Runtime.instance [as handler] (/var/task/node_modules/@middy/core/index.js:141:12)",
        "    at Runtime.handleOnce (/var/runtime/Runtime.js:66:25)"
    ]
}
```

The error is on a function whose handler looks like this:
```
module.exports.handler = wrap(async (event, context) => {
  const restaurants = await getRestaurants(process.env.defaultResults)
  const response = {
    statusCode: 200,
    body: JSON.stringify(restaurants)
  }

  return response
}).use(ssm({
  cache: true,
  cacheExpiryInMillis: 5 * 60 * 1000, // 5 mins
  names: {
    config: `/${serviceName}/${stage}/get-restaurants/config`
  },
  onChange: () => {
    const config = JSON.parse(process.env.config)
    process.env.defaultResults = config.defaultResults
  }  
}))
```

I haven't had the time to look into what's wrong with it exactly yet, but thought I'd give you a heads up that something's broken between v1.5.2 and v2.0.0-alpha4

**Comment by willfarrell** [comment]
Thanks for reporting. In v2.x there was a breaking change to all middleware that reach out to AWS services to make them consistent with each other and enable new functionality. I've documented all breaking changes in the [UPGRADE.md](https://github.com/middyjs/middy/blob/release/2.x/docs/UPGRADE.md)

The configuration you're showing is not compatible with v2. Something like this should work:

```js
.use(ssm({
  cacheExpiry: 5 * 60 * 1000, // 5 mins
  fetchData:{
    config: `/${serviceName}/${stage}/get-restaurants/config`
  },
  setToEnv: true,
  onChange: () => {
    const config = JSON.parse(process.env.config)
    process.env.defaultResults = config.defaultResults
  }  
}))
```
 See https://github.com/middyjs/middy/blob/release/2.x/packages/ssm/README.md#options for more new options.

**Comment by willfarrell** [comment]
I've been reflecting on this today. To allow effective onChange detection to meet your use case, there are two option I can think of.

1. Pass `hander` to the `onChange` function and add `await`.
```js
middy(handler)
  .use(ssm({
    ...
    onChange: async (handler) => {
      const value = await getInternal('config.defaults', handler)
    }
  }))
```
To me this kinda feels like a hack.

2. What if there is no `onChange` function on any middlewares ... What if `handler.internal` was a es6 `Proxy` and you could add any kind of catch you want to the setter/getter? 

```.js
middy(originalHandler)
  .before((handler) => {
    handler.internal = new Proxy({}, {
      set: (obj, prop, value) => {
        if(prop === 'ssmInernalKey') {
          // do ssm onChange just for the one key
        }
        obj[prop] = value
        return true
      }
    }
  })
  .use(ssm({
    ...
  }))
```
I still need to test and sleep on this, but 

@theburningmonk @lmammino What are your thoughts on this? Is there another solution?

**Comment by lmammino** [comment]
I might be missing some important architectural detail right now, but I think the `onChange` syntax still feels like the easiest mental model for me...

Maybe we should just change the semantic of it and make clear that you won't/shouldn't be able to access env vars.

Could we implement something like this?

```javascript
middy(handler)
  .use(ssm({
    ...
    onChange: async (data) => {
      // `data` has the new data
    }
  }))
```

As a side note, I am realising that in [the docs for v2](https://github.com/middyjs/middy/tree/release/2.x/packages/ssm) we only show how to fetch the data but not how to read it. Definitely room for a small improvement here.

**Comment by willfarrell** [comment]
Did some sleeping on this. Option 2 I suggested will technically will trigger every call, so that's no good anyways.  The current implementation doesn't properly trigger on change now. So I'll have to address that first and add in the missing test. While I resolve that, I'll see if another solution pops up. I agree calling the an onChange function with the resolved data if the easiest mentally, but could incur a performance hit in some use case (ie requesting >10 values) compared to option 1 I suggested above. I've also been thinking about needing the documentation to so the best way to chain middlewares that need to know when an upstream middleware requests a new value as well.
